File: src/main/java/nl/crosscode/x509certbuddy/decoder/CertRetriever.java
package nl.crosscode.x509certbuddy.decoder;

import com.intellij.openapi.editor.Editor;

import java.io.ByteArrayInputStream;
import java.io.IOException;
import java.io.InputStream;
import java.security.cert.CertificateException;
import java.security.cert.CertificateFactory;
import java.security.cert.X509Certificate;
import java.util.ArrayList;
import java.util.List;
import java.util.stream.Collectors;

public class CertRetriever {
    private final CertificateFactory certFactory;
    private final Editor editor;

    public CertRetriever(Editor editor) throws CertificateException {
        this.editor = editor;
        certFactory = CertificateFactory.getInstance("X.509");
    }

    public List<RetrievedCert> retrievedCerts(byte[] data) throws CertificateException {
        if (data==null||data.length<64) return List.of();
        if (data[0]==0x30) {
            try (ByteArrayInputStream bais = new ByteArrayInputStream(data)) {
                return CertificateFactory.getInstance("X.509").generateCertificates(bais).stream().map(c->new RetrievedCert(null,0,(X509Certificate)c)).collect(Collectors.toList());
            } catch (IOException | CertificateException e) {
            }
        }
        return retrieveCerts(new String(data));
    }

    public List<RetrievedCert> retrieveCerts(String text) throws CertificateException {

        List<RetrievedCert> certificates = new ArrayList<>();
        getCertsFromText(text, certificates);
        return certificates;
    }

    private void getCertsFromText(String text, List<RetrievedCert> certificates) {
        for (PotentialCert potentialCert : findPotentialCerts(text)) {
            try {
                for (X509Certificate cert : certFromBytes(potentialCert.getPotentialCert())) {
                    certificates.add(new RetrievedCert(editor, potentialCert.getOffset(), cert));
                }
            } catch (Exception e) {} // Ignoring it for now due to the brute force nature of cert finding.
        }
    }

    private List<X509Certificate> certFromBytes(byte[] bytes) throws CertificateException {
        InputStream in = new ByteArrayInputStream(bytes);
        return certFactory.generateCertificates(in).stream().map(c->(X509Certificate)c).collect(Collectors.toList());
    }

    private List<PotentialCert> findPotentialCerts(String text) {
        List<PotentialCert> potentialCerts = new ArrayList<>();
        List<Decoder> decoders = new ArrayList<Decoder>();
        int offset = -1;
        for (char c : text.toCharArray()) {
            offset++;
            if (c=='M') {
                decoders.add(new Decoder(offset));
            }
            for (Decoder decoder : decoders) {
                if (decoder.add(c)) { // decoder is certainly done
                    decodeToPotentialCerts(potentialCerts, decoder);
                }
            }
            decoders = decoders.stream().filter(decoder -> !decoder.isDone()).collect(Collectors.toList());
        }
        for (Decoder decoder : decoders) {
            decodeToPotentialCerts(potentialCerts, decoder);
        }
        return potentialCerts;
    }

    private static void decodeToPotentialCerts(List<PotentialCert> potentialCerts, Decoder decoder) {
        SeqDerReader seqDerReader = new SeqDerReader();
        try {
            byte[] data = decoder.tryDecode();
            for (byte b : data) {
                seqDerReader.read(b);
            }
            seqDerReader.setEof();
            if (!seqDerReader.isError()) {
                potentialCerts.add(new PotentialCert(seqDerReader.getResult(),decoder.getOriginalOffset()));
            }
        } catch (OutOfMemoryError | Exception e) {}
    }
}

File: src/main/kotlin/nl/crosscode/x509certbuddy/decoder/CertRetriever.kt
package nl.crosscode.x509certbuddy.decoder

import com.intellij.openapi.editor.Editor
import java.io.ByteArrayInputStream
import java.io.IOException
import java.io.InputStream
import java.security.cert.Certificate
import java.security.cert.CertificateException
import java.security.cert.CertificateFactory
import java.security.cert.X509Certificate
import java.util.stream.Collectors

class CertRetriever(private val editor: Editor?) {
    private val certFactory: CertificateFactory = CertificateFactory.getInstance("X.509")

    @Throws(CertificateException::class)
    fun retrieveCerts(data: ByteArray): List<RetrievedCert> {
        if (data.size < 64) return listOf()
        if (data[0].toInt() == 0x30) {
            try {
                ByteArrayInputStream(data).use { bais ->
                    return CertificateFactory.getInstance("X.509").generateCertificates(bais).stream()
                        .map { c: Certificate? -> RetrievedCert(null, 0, c as X509Certificate) }
                        .collect(Collectors.toList())
                }
            } catch (e: IOException) {
            } catch (e: CertificateException) {
            }
        }
        return retrieveCerts(String(data))
    }

    @Throws(CertificateException::class)
    fun retrieveCerts(text: String): List<RetrievedCert> {
        val certificates: MutableList<RetrievedCert> = ArrayList()
        getCertsFromText(text, certificates)
        return certificates
    }

    private fun getCertsFromText(text: String, certificates: MutableList<RetrievedCert>) {
        for (potentialCert in findPotentialCerts(text)) {
            try {
                for (cert in certFromBytes(potentialCert.potentialCert)) {
                    certificates.add(RetrievedCert(editor, potentialCert.offset, cert))
                }
            } catch (e: Exception) {
            } // Ignoring it for now due to the brute force nature of cert finding.
        }
    }

    @Throws(CertificateException::class)
    private fun certFromBytes(bytes: ByteArray?): List<X509Certificate> {
        val `in`: InputStream = ByteArrayInputStream(bytes)
        return certFactory.generateCertificates(`in`).stream().filter { c -> c is X509Certificate } .map { c: Certificate -> c as X509Certificate }
            .collect(Collectors.toList())
    }

    private fun findPotentialCerts(text: String): List<PotentialCert> {
        val blockingDecoders: MutableList<Decoder> = ArrayList()
        val potentialCerts: MutableList<PotentialCert> = ArrayList()
        var decoders: MutableList<Decoder> = ArrayList()
        var offset = -1
        for (c in text.toCharArray()) {
            offset++
            if (c == 'M') {
                decoders.add(Decoder(offset))
            }
            for (decoder in decoders) {
                if (decoder.add(c)) { // decoder is certainly done
                    if (blockingDecoders.stream()
                            .noneMatch { blockingDecoder: Decoder -> blockingDecoder.isOffsetInsideRange(decoder.originalOffset) }
                    ) { // decoder is not working inside another decoders range which already succeeded.
                        blockingDecoders.add(decoder)
                        decodeToPotentialCerts(potentialCerts, decoder)
                    }
                }
            }
            decoders = decoders.stream().filter { decoder: Decoder -> !decoder.isDone }
                .collect(Collectors.toList())
        }
        for (decoder in decoders) {
            decodeToPotentialCerts(potentialCerts, decoder)
        }
        return potentialCerts
    }

    companion object {
        private fun decodeToPotentialCerts(potentialCerts: MutableList<PotentialCert>, decoder: Decoder) {
            val seqDerReader = SeqDerReader()
            try {
                val data = decoder.tryDecode()
                for (b in data) {
                    seqDerReader.read(b)
                }
                seqDerReader.setEof()
                if (!seqDerReader.isError) {
                    potentialCerts.add(PotentialCert(seqDerReader.getResult(), decoder.originalOffset))
                }
            } catch (e: OutOfMemoryError) {
            } catch (e: Exception) {
            }
        }
    }
}
